src/errors/join.go

```diff
@@ -17,24 +17,33 @@ import (
 //
 // A non-nil error returned by Join implements the Unwrap() []error method.
 func Join(errs ...error) error {
-	n := 0
-	for _, err := range errs {
-		if err != nil {
-			n++
+	allErrs := make([]error, 0, len(errs))
```

Comment:
This might be one of those situations where optimizing for the common use case makes sense. Given "most" error joining is of the form `merr = errors.Join(merr, err)`, we can optimize for the single-entry error:

```go
switch len(errs) {
  case 0:
    return nil
  case 1:
    return errs[0]
  case 2:
    if errs[0] == nil {
      return errs[1]
    } else if errs[1] == nil {
      return errs[0]
    } else {
      return &joinError{
        errs: errs, 
      }
    }
  default:
    // as below
}
```